ehr/src/org/labkey/ehr/utils/TriggerScriptHelper.java

```diff
@@ -1215,30 +1237,106 @@ private String checkOtherDrawsQuantity(String id, Date date, String rowObjectId,
             {
                 overages.add(bloodNextInterval);
             }
+            else if (getThresholdCheck())
+            {
+                //only report about problematic bloods nearing the limit in the current transaction
+                //because allBloods contains everything and we want to distinguish them
+                if (blood1.getIsInTransaction())
+                {
+                    double maxAllowableThreshold = maxAllowable - getThresholdLimit();
+                    if (bloodNextInterval > maxAllowableThreshold)
+                    {
+                        closeToThreshold.add(bloodNextInterval);
+                    }
+                }
+            }
         }
 
-        //always report the most severe overage
-        if (!overages.isEmpty())
+        if (!overages.isEmpty() || !closeToThreshold.isEmpty())
+        {
+            StringBuilder errorMsgBuilder = new StringBuilder();
+
+            //always report the most severe overage
+            if (!overages.isEmpty())
+            {
+                errorMsgBuilder.append("Blood volume of ");
+                errorMsgBuilder.append(rowQuantity);
+                errorMsgBuilder.append(" (");
+                errorMsgBuilder.append(overages.descendingSet().iterator().next());
+                errorMsgBuilder.append(" over ");
+                errorMsgBuilder.append(interval);
+                errorMsgBuilder.append(" days) exceeds the allowable volume of ");
+                errorMsgBuilder.append(maxAllowable);
+                errorMsgBuilder.append(" mL (weight: ");
+                errorMsgBuilder.append(weight);
+                errorMsgBuilder.append(" kg).\n");
+            }
+
+            //report all the ones that are close to limit
+            for (double closeValue : closeToThreshold.descendingSet())
```

Comment:
Would you ever want to report multiple messages? Should you report the most severe, like above? Or try to distill multiple closeToThreshold values into one string message?

Also: should we differentiate between actual overages (which seems clearly like an error), and close-to-overage (which seems like maybe it should be a warning? While we just pass a basic string to the client, maybe we could modify this to prefix with "WARNING:", and have the JS side test for this, and convert this to a warning not error?

Reply:
I believe the users would like all of the messages, what do you mean by distill down?

For the error levels, we are doing something like that, but about one level less, meaning the close-to-overages are INFO level and the over the limits are WARN. We are controlling this error severity level on the JS side like you mention, [like this](https://github.com/LabKey/wnprc-modules/blob/893498276fb0436accbf5486b1353c892135cbcf/WNPRC_EHR/resources/scripts/wnprc_ehr/wnprc_triggers.js#L809) but by using the prefix "Limit" (instead of using a prefix like WARNING or ERROR, to not confuse later if those error severity levels are set differently for each center), and then using the prefix to set the error the threshold to INFO (only if there's no overages), while keeping the ones that are overages as WARNING. This lets me use the same string without refactoring a bunch of things. Blood draw requests never let us submit when there's a WARNING, unless it's a privileged user, afaik.

Reply:
One more comment: when you add javadoc to the setThreshold() method (and I might rename that to make it more descriptive), can you please mention that true overages and near-overages will be returned as one array, but the consumer than separate them based on the message (or something to that effect). We should at least put some effort into making this halfway reasonable for a future developer to decipher how to use it. Maybe even put a link to your WNPRC github code in the javdoc?

To the other question about distilling errors. At least for actual overages, it is possible for a given blood record to cause multiple problems with different volumes, assuming there are multiple future draws in question. That existing code collapses what would be multiple redundant errors into a single message and reports the max volume of the overage. 

Unless I"m missing something, I am questioning whether this same pattern makes sense for near-overages. If a given blood record would put the limit within 2mL based on some record and also 1mL based on some other combination of records, is there a reason the user needs to see two different messages that only differ by their volume?